File: examples/grab.rs
//! Full pipeline smoke test: discover/connect, read features, stream frames.
//! `cargo run --example grab <camera-ip> [n-frames]`

use std::time::Duration;

use telegenic::{GenICamera, StreamConfig};

fn main() {
    tracing_subscriber::fmt::init();
    let mut args = std::env::args().skip(1);
    let ip: std::net::IpAddr = args
        .next()
        .expect("usage: grab <camera-ip> [n-frames]")
        .parse()
        .expect("camera ip");
    let n_frames: usize = args.next().map_or(10, |s| s.parse().expect("frame count"));

    let mut cam = GenICamera::new(ip);
    cam.connect().expect("connect");
    {
        let info = cam.transport().device_info().expect("device info");
        println!(
            "{} {} (serial {})",
            info.manufacturer, info.model, info.serial
        );
    }

    for feature in ["Width", "Height", "PayloadSize"] {
        match cam.get_integer(feature) {
            Ok(v) => println!("  {feature} = {v}"),
            Err(e) => println!("  {feature}: {e}"),
        }
    }
    for feature in ["PixelFormat", "AcquisitionMode", "TriggerMode"] {
        match cam.get_enum(feature) {
            Ok(v) => println!("  {feature} = {v}"),
            Err(e) => println!("  {feature}: {e}"),
        }
    }
    match cam.get_float("ExposureTime") {
        Ok(v) => println!("  ExposureTime = {v}"),
        Err(e) => println!("  ExposureTime: {e}"),
    }

    let stream = cam
        .start_acquisition(StreamConfig::new(0))
        .expect("start acquisition");
    println!(
        "streaming to {} with packet size {}",
        stream.local_addr(),
        stream.packet_size()
    );
    let frames = stream.subscribe(16);

    let mut received = 0usize;
    let deadline = std::time::Instant::now() + Duration::from_secs(15);
    while received < n_frames && std::time::Instant::now() < deadline {
        let Some(frame) = frames.wait_for(Duration::from_millis(500)) else {
            continue;
        };
        received += 1;
        let data = frame.data();
        let sum: u64 = data.iter().map(|&b| u64::from(b)).sum();
        println!(
            "frame {:>4}  {:?}  {}x{} {}  {} bytes  ts {} ns  mean {:.1}",
            frame.frame_id,
            frame.status,
            frame.width,
            frame.height,
            frame.pixel_format,
            frame.received_size,
            frame.timestamp_ns,
            sum as f64 / data.len().max(1) as f64,
        );
    }

    cam.stop_acquisition().expect("stop acquisition");
    println!("stream stats: {:#?}", stream.stats());
    println!("link stats: {:?}", cam.transport().stats().expect("stats"));
    assert!(received > 0, "no frames received");
}

File: examples/grab.py
"""Full pipeline smoke test: connect, read features, stream frames.

python examples/grab.py <camera-ip> [n-frames]
"""

import sys
import time

import telegenic


def main() -> None:
    if len(sys.argv) < 2:
        sys.exit("usage: grab.py <camera-ip> [n-frames]")
    ip = sys.argv[1]
    n_frames = int(sys.argv[2]) if len(sys.argv) > 2 else 10

    cam = telegenic.Camera(ip)
    cam.connect()
    info = cam.device_info()
    print(f"{info.manufacturer} {info.model} (serial {info.serial})")

    for feature in ["Width", "Height", "PayloadSize"]:
        try:
            print(f"  {feature} = {cam.get_integer(feature)}")
        except telegenic.GenicamError as e:
            print(f"  {feature}: {e}")
    for feature in ["PixelFormat", "AcquisitionMode", "TriggerMode"]:
        try:
            print(f"  {feature} = {cam.get_enum(feature)}")
        except telegenic.GenicamError as e:
            print(f"  {feature}: {e}")

    with cam.start_acquisition() as acq:
        print(f"streaming to {acq.local_addr()} with packet size {acq.packet_size()}")
        received = 0
        deadline = time.monotonic() + 15.0
        while received < n_frames and time.monotonic() < deadline:
            frame = acq.wait_for(timeout=0.5)
            if frame is None:
                continue
            received += 1
            data = frame.data()
            mean = sum(data) / max(len(data), 1)
            print(
                f"frame {frame.frame_id:>4}  {frame.status}  "
                f"{frame.width}x{frame.height} {frame.pixel_format}  "
                f"{frame.received_size} bytes  ts {frame.timestamp_ns} ns  "
                f"mean {mean:.1f}"
            )
        stats = acq.stats()
        print(
            f"stream stats: {stats.completed_frames} complete, "
            f"{stats.failed_frames} failed, {stats.missing_packets} missing packets, "
            f"{stats.resend_requests} resends, {stats.underruns} underruns"
        )

    print(f"link stats: {cam.link_stats()}")
    cam.disconnect()
    assert received > 0, "no frames received"


if __name__ == "__main__":
    main()
